aips/object-code-deployment-auth-ref.md

```diff
@@ -0,0 +1,167 @@
+---
+aip:
+title: Add signer capability support for object code deployment
+author: mshakeg
+discussions-to:
+Status: Draft
+type: Standard (Framework)
+created: 02/17/2024
+---
+
+# AIP-X - Add signer capability support for object code deployment
+
+## Summary
+
+When publishing a package under an object, modules cannot access their object signer after initialization, making it impossible to manage funds sent to the object's primary fungible store. This AIP proposes adding AuthRef support to enable modules to securely access their object signer during initialization and store it for later use, similar to how modules published under resource accounts can retrieve and store their SignerCapability.
+
+The solution introduces an AuthRef abstraction that can only be generated during module initialization, maintaining security while enabling proper fund management capabilities for object-published modules.
+
+### Out of scope
+
+- Changes to how packages are published under objects
+- Changes to existing package upgrade mechanisms
+- Additional permissioned functions for object management
+- Changes to modules already published under objects (cannot retroactively gain AuthRef access)
+
+## High-level Overview
+
+The proposal adds AuthRef support to the `object_code_deployment` module, enabling modules published under objects to generate an AuthRef during initialization that can later be used to generate the object's signer. This brings object-based publishing closer to parity with resource account-based publishing, where modules can already retrieve and store their SignerCapability during initialization.
+
+The AuthRef serves as a capability, proving that a module has the right to generate its object's signer. It can only be created during module initialization when the object signer is naturally available, following Move's capability-based security model.
```

Comment:
Thanks for starting this AIP! Can I ask what’s the rationale for adding this an auth ref that would require explicit init module template code vs just adding a function to allow arbitrary access to signer as long as that function is called from the same address (or other similar options with more safeguards)? The approach in AuthRef seems longer and less convenient to use for developers. They'd have copy paste code for init module (similar to https://github.com/aptos-labs/aptos-core/tree/main/aptos-move/move-examples/package_manager) to ensure they don't lose access later. Having a more direct way would be better dev ex as people don’t need to remember and existing code can get auto access as long as they’re upgradable

Reply:
An issue with deploying to resource accounts is that if you don't deploy `package_manager` along with your contracts, you forfeit access to the `signer`. With the `AuthRef` having to be created during construction, it faces a similar problem where if you do not have this logic in your contract , you lose access to the `signer`.

One option could be to have the owner of the code object generate `AuthRef`, something similar to the collection owner having access to the collection object `signer` [here](https://github.com/aptos-labs/token-minter/blob/main/token-minter/sources/modules/collection_components.move#L249).
This means if they forget, they can call a function to generate the `AuthRef` and store it at a later stage.

Reply:
@JohnChangUK when a pkg is upgraded is `init_module` called again? Or is it only called once at the time of initial deployment? If it's the latter then I agree it could be useful to extend the `generate_auth_ref` to also accept the signer of the owner of the pkg object, however if it's the former then the pkg could be upgraded to have `init_module` called again to have the module call `generate_auth_ref` in this upgraded `init_module`?

Reply:
Upgrades do not invoke the init module function from what I remember.

Reply:
@JohnChangUK ok, then I agree that `generate_auth_ref` should also allow for the pkg object owner to generate the `AuthRef`

Reply:
I'm thinking there doesn't need to be an "owner" defined. Basically there will be this function

```
public fun package_signer(): signer {}
```

that's only callable from the same address